### PR TITLE
[gemma3] fix bidirectional image mask

### LINKED ISSUE #39389: Gemma3 bidirectional mask for image tokens isn't reaching attention forward
### System Info

Transformers version: 4.53.2
Ubuntu x86, Nvidia H100

- I've noticed previous [issue](https://github.com/huggingface/transformers/issues/38053) about bidirectional mask for image tokens and MRs fixing it merged[[1](https://github.com/huggingface/transformers/pull/38080)][[2](https://github.com/huggingface/transformers/pull/38295)].
- I was looking at the attention mask being passed into attention forward at prefill for a single image and noticed that the mask is purely causal i.e. no bidirectional mask for image tokens.
- I'm afraid there might be a plumbing issue. Can someone please take a look?

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Script to reproduce:
```
from transformers import AutoProcessor, Gemma3ForConditionalGeneration

model_dir = "path/to/gemma-3-4b-it/"
model = Gemma3ForConditionalGeneration.from_pretrained(model_dir)
processor = AutoProcessor.from_pretrained(model_dir)
messages = [
    {
        "role": "system",
        "content": [
            {"type": "text", "text": "You are a helpful assistant."}
        ]
    },
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg"},
            {"type": "text", "text": "Describe this image in detail."},
        ]
    },
]

inputs = processor.apply_chat_template(messages, tokenize=True, return_dict=True, return_tensors="pt", add_generation_prompt=True)
generate_ids = model.generate(**inputs, do_sample=False, num_beams=1, top_p=None, top_k=None, max_new_tokens=2)
outputs = processor.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
print(outputs)
```

### Expected behavior

Mask should be bidirectional for images as shown in this MR description:
Reference 1: https://github.com/huggingface/transformers/issues/38053
Reference 2: https://github.com/huggingface/transformers/pull/38295

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/39389. There were two issues. First, using `getattr` on dict will never look for a dict key thus we were skipping on `token_type_ids_mask`. 

Second, `kv_idx` can go beyond `token_type_ids` length because it includes the new tokens to be generated. Using a simple check on `ids.shape` fails in vmap due to dynamic control flow. One option would be to create dummy padded token type ids using `kv_idx` but I think a simple try-except is also fine

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: gemma3

**Comment by zucchini-nlp** [comment]
run-slow: gemma3

**Comment by zucchini-nlp** [comment]
Just realized the `is_same_image` doesn't work correctly and unmasks bidirectional mask to future images. Working on it, don't review yet

**Comment by zucchini-nlp** [comment]
run-slow: gemma3

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 1]
```diff
@@ -767,10 +771,18 @@ def token_type_ids_mask_function(token_type_ids: Optional[torch.Tensor], tokens_
...
+        # Since vmap doesn't support `if statement` we workaround it with `torch.where`
+        safe_idx = torch.where(kv_idx < token_type_ids.shape[1], kv_idx, 0)
+        token_type_ids_at_kv_idx = token_type_ids[batch_idx, safe_idx]
+        token_type_ids_at_kv_idx = torch.where(kv_idx < token_type_ids.shape[1], token_type_ids_at_kv_idx, 0)
+
+        image_group_ids_at_kv_idx = image_group_ids[batch_idx, safe_idx]
+        image_group_ids_at_kv_idx = torch.where(kv_idx < image_group_ids.shape[1], image_group_ids_at_kv_idx, -1)
+
+        is_image_block = (token_type_ids[batch_idx, q_idx] == 1) & (token_type_ids_at_kv_idx == 1)
+        same_image_block = image_group_ids[batch_idx, q_idx] == image_group_ids_at_kv_idx
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma3/modeling_gemma3.py:783]: we could use only `token_type_ids` but since `vmap` doesn't support slicing, it always fails with an error. I tried out with different torch ops, none of them worked. Passing `image_group_ids` was the most straightforward woraround

**Comment by zucchini-nlp** [comment]
Can be reviewed now, slow tests are failing and I'll fix them tomorrow. No idea how to get `read_token` from inside runners, will ask Yih-Dar

Edit: yeah, we need to login with our own access tokens after ssh-ing

**Comment by zucchini-nlp** [comment]
run-slow: gemma3

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/gemma3/modeling_gemma3.py:783]: looks nice!

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 2]
```diff
@@ -935,8 +947,15 @@ def forward(
             }
             if token_type_ids is not None and inputs_embeds.shape[1] != 1:
                 # We need to pass an additional mask function to account for token type ids, and it needs to be an `or`
+
+                # First find where a new image block starts: 1 if image and previous not image
+                # The images cannot attend to future images, but can attend to all prev images and to itself bidirectionally
+                is_image = (token_type_ids == 1).to(cache_position.device)
+                new_image_start = is_image & ~nn.functional.pad(is_image, (1, 0), value=0)[:, :-1]
+                image_group_ids = torch.cumsum(new_image_start.int(), dim=1) - 1
+                image_group_ids = torch.where(is_image, image_group_ids, torch.full_like(token_type_ids, -1))
```
Comment (ArthurZucker) [inline src/transformers/models/gemma3/modeling_gemma3.py:954]: should maybe put this in the `token_type_ids_mask_function`?

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 3]
```diff
@@ -1207,8 +1226,15 @@ def create_masks_for_generate(
         # Add the token type ids mask for generate as well
         if token_type_ids is not None and input_embeds.shape[1] != 1:
             # We need to pass an additional mask function to account for token type ids, and it needs to be an `or`
+
+            # First find where a new image block starts: 1 if image and previous not image
+            # The images cannot attend to future images, but can attend to all prev images and to itself bidirectionally
+            is_image = (token_type_ids == 1).to(cache_position.device)
+            new_image_start = is_image & ~nn.functional.pad(is_image, (1, 0), value=0)[:, :-1]
+            image_group_ids = torch.cumsum(new_image_start.int(), dim=1) - 1
+            image_group_ids = torch.where(is_image, image_group_ids, torch.full_like(token_type_ids, -1))
```
Comment (ArthurZucker) [inline src/transformers/models/gemma3/modeling_gemma3.py:1233]: ah you said we can't for compile?

**Review by ArthurZucker — APPROVED** [review]
Can we have a fast test for this? 🤗

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:1233]: yeah, this is super verbose because of compile and vmap's own restrictions

**Comment by zucchini-nlp** [comment]
Actually the slow tests should fail whenever it breaks, prob we didn't notice it because slow tests were failing for another reason already. I will try to make a fast test to check attentions
